File: reference/src/appearance.ts
// 見た目の決定（色違い・汚れ・破損・保存状態）。docs/04。すべて純粋。
// 離散判定（dirt/damage）は整数しきい値、連続出力（colorMod/preservation）は f64（docs/08 8.4）。
import { uniform, u32, pickWeighted } from "./hash.ts";
import { TAGS, DIRT_LAYERS, DAMAGE_CUTOFFS_U32 } from "./constants.ts";
import type { ColorMod, Damage } from "./types.ts";

/** 色違い HSV 補正（連続出力＝f64。レア度と独立, docs/04 4.1）。 */
export function colorMod(seed: Uint8Array): ColorMod {
  return {
    hShift: uniform(seed, TAGS.color) * 360 - 180, // -180..+180
    sMul: 0.6 + uniform(seed, TAGS.colorS) * 0.8, // 0.6..1.4
    vMul: 0.7 + uniform(seed, TAGS.colorV) * 0.6, // 0.7..1.3
  };
}

/** 汚れレイヤーID（離散・整数 pickWeighted, docs/04 4.2）。 */
export function dirtLayerId(seed: Uint8Array): string {
  return pickWeighted(seed, TAGS.dirt, DIRT_LAYERS);
}

/** 破損 ON/OFF（離散・整数しきい値, docs/04 4.3）。各種が独立ストリーム。 */
export function damage(seed: Uint8Array): Damage {
  return {
    chip: u32(seed, TAGS.damageChip) < DAMAGE_CUTOFFS_U32.chip,
    crack: u32(seed, TAGS.damageCrack) < DAMAGE_CUTOFFS_U32.crack,
    wear: u32(seed, TAGS.damageWear) < DAMAGE_CUTOFFS_U32.wear,
  };
}

/**
 * 保存状態スコア 0..1（docs/04 4.4）。
 * preserve ストリーム自体はレア度と独立。finalRarity で弱い加算バイアスのみ。
 */
export function preservationScore(seed: Uint8Array, finalRarity: number): number {
  const base = uniform(seed, TAGS.preserve);
  const bias = ((finalRarity - 1) / 12) * 0.25; // 高レアほど最大 +0.25
  return Math.max(0, Math.min(1, base + bias));
}

// --- レイヤー強度（合成時の不透明度, docs/04 4.3.1）。見た目のみで決定論入力は汚さない。 ---
export function dirtStrength(s: number): number {
  return Math.max(0.1, Math.min(0.9, 0.9 - 0.7 * s));
}
export function damageStrength(s: number): number {
  return Math.max(0.15, Math.min(0.9, 0.9 - 0.6 * s));
}

File: rust/qrac-core/src/appearance.rs
//! 見た目（色違い・汚れ・破損・保存状態）。docs/04。
//! 離散（dirt/damage）は整数しきい値、連続（colorMod/preservation）は f64（docs/08 8.4）。
use crate::constants::*;
use crate::hash::{pick_weighted, u32v, uniform};
use crate::types::{ColorMod, Damage};

/// 色違い HSV 補正（連続・f64）。
pub fn color_mod(seed: &[u8]) -> ColorMod {
    ColorMod {
        h_shift: uniform(seed, TAG_COLOR) * 360.0 - 180.0,
        s_mul: 0.6 + uniform(seed, TAG_COLOR_S) * 0.8,
        v_mul: 0.7 + uniform(seed, TAG_COLOR_V) * 0.6,
    }
}

/// 汚れレイヤーID（離散・整数 pick_weighted）。
pub fn dirt_layer_id(seed: &[u8]) -> String {
    pick_weighted(seed, TAG_DIRT, &DIRT_LAYERS)
}

/// 破損 ON/OFF（離散・整数しきい値）。
pub fn damage(seed: &[u8]) -> Damage {
    Damage {
        chip: u32v(seed, TAG_DAMAGE_CHIP, 0) < DAMAGE_CUTOFF_CHIP,
        crack: u32v(seed, TAG_DAMAGE_CRACK, 0) < DAMAGE_CUTOFF_CRACK,
        wear: u32v(seed, TAG_DAMAGE_WEAR, 0) < DAMAGE_CUTOFF_WEAR,
    }
}

/// 保存状態スコア 0..1（連続・f64）。レア度で弱い加算バイアス。
pub fn preservation_score(seed: &[u8], final_rarity: u32) -> f64 {
    let base = uniform(seed, TAG_PRESERVE);
    let bias = ((final_rarity as f64 - 1.0) / 12.0) * 0.25;
    (base + bias).clamp(0.0, 1.0)
}
